backends/p4tools/testgen/targets/bmv2/p4_asserts_parser.cpp

```diff
@@ -0,0 +1,663 @@
+/// Functions for multiple comparison kind
+bool Token::is_one_of(Token::Kind k1, Token::Kind k2) const noexcept { return is(k1) || is(k2); }
+template <typename... Ts>
+bool Token::is_one_of(Token::Kind k1, Token::Kind k2, Ts... ks) const noexcept {
+    return is(k1) || is_one_of(k2, ks...);
+}
+/// The function to get lexeme from token
+std::string_view Token::lexeme() const noexcept { return m_lexeme; }
+/// The function to replace the token lexeme with another lexeme
+void Token::lexeme(std::string_view lexeme) noexcept { m_lexeme = lexeme; }
+
+/// Function to get the current character
+char Lexer::peek() const noexcept { return *m_beg; }
+/// The function advances the iterator one index back and returns the character corresponding to the
+/// position of the iterator
+char Lexer::prev() noexcept { return *m_beg--; }
+/// The function advances the iterator one index forward and returns the character corresponding to
+/// the position of the iterator
+char Lexer::get() noexcept { return *m_beg++; }
+
+bool isSpace(char c) noexcept {
+    switch (c) {
+        case ' ':
+        case '\t':
+        case '\r':
+            return true;
+        default:
+            return false;
+    }
+}
+
+std::ostream& operator<<(std::ostream& os, const Token::Kind& kind) {
+    static const char* const NAMES[]{
+        "Priority",    "Text",           "True",         "False",       "LineStatementClose",
+        "Id",          "Number",         "Minus",        "Plus",        "Dot",
+        "FieldAcces",  "MetadataAccess", "LeftParen",    "RightParen",  "Equal",
+        "NotEqual",    "GreaterThan",    "GreaterEqual", "LessThan",    "LessEqual",
+        "LNot",        "Colon",          "Semicolon",    "Conjunction", "Disjunction",
+        "Implication", "Slash",          "Percent",      "Shr",         "Shl",
+        "Mul",         "Comment",        "Unknown",      "EndString",   "End",
+    };
+    return os << NAMES[static_cast<int>(kind)];
+}
+
+/// The function combines IR expressions separated by the "or" ,"and" operators
+/// and also by the implication which is indicated by "(tmp" ,inside the algorithm,
+/// for the convenience of determining the order of expressions. At the output, we get one
+/// expression. For example, at the input we have a vector of expressions: [IR::Expression, IR::LOr
+/// with name "(tmp", IR::Expression, IR::LAnd, IR::Expression] The result will be an IR::Expression
+/// equal to !IR::Expression || (IR::Expression && IR::Expression)
+const IR::Expression* makeSingleExpr(std::vector<const IR::Expression*> input) {
+    const IR::Expression* expr = nullptr;
+    for (uint64_t i = 0; i < input.size(); i++) {
+        if (const auto* lOr = input[i]->to<IR::LOr>()) {
+            if (lOr->right->toString() == "(tmp") {
+                if (i + 1 == input.size()) {
+                    break;
+                }
+                std::vector<const IR::Expression*> tmp;
+                uint64_t idx = i;
+                i++;
```

Comment:
There is still a bunch of i++ in this code. They should not be necessary.

Reply:
I do not think this has been fixed yet.

Reply:
Look, I changed the algorithm now removed this part